### PR TITLE
Add a ping service

### PR DESCRIPTION
**Pull request by ericxtang**
Add the ability to ping another node.  This can be used for testing connectivity.

### CODE REVIEW

File: basic_out_stream.go [hunk 1]
```diff
@@ -92,13 +92,13 @@ func (bs *BasicOutStream) encodeAndFlush(n interface{}) error {
 	err := bs.enc.Encode(n)
 	if err != nil {
 		glog.Errorf("send message encode error for peer %v: %v", peer.IDHexEncode(bs.Stream.Conn().RemotePeer()), err)
-		return ErrOutStream
+		return err
```
Comment (j0sh) [inline basic_out_stream.go:95]: Curious, why this change? In particular, is there a reason we were returning `ErrOutStream` specifically before?

File: data.go [hunk 2]
```diff
@@ -138,6 +140,11 @@ func (m Msg) MarshalJSON() ([]byte, error) {
 		if err != nil {
 			return nil, fmt.Errorf("Failed to marshal NodeStatusDataMsg: %v", err)
 		}
+	case string:
```
Comment (j0sh) [inline data.go]: Since all the other message types have their own payloads, should we do the same here for consistency and a better chance at forward compatibility?

**Review by j0sh — CHANGES_REQUESTED** [review]
Since libp2p messages are unreliable, the ping should include some sort of nonce in the payload that the pong mirrors. This gives us much more visibility into network conditions -- we can match up specific ping/pong messages, discover any loss, calculate round-trip latency, etc.

There's a question of when / how Ping should be invoked. I see that it's meant to (manually?) check the bootnodes within goclient. Thats OK, but the usefulness there seems limited to a connectivity check + reconnect attempt.

One thought is that pings should be used as a keepalive internal to basicnet; that is, sent out periodically, for each connection (regardless of who initiated the connection, so it'd be bidirectional).

That brings up the question of what to do if this keepalive detects a non-transient broken connection, but that's perhaps a topic for a separate discussion.

File: basic_out_stream.go (hunk 1 above)
Comment (ericxtang) [reply basic_out_stream.go:95]: I thought `ErrOutStream` was hiding the real error

**Comment by ericxtang** [comment]
@j0sh why are libp2p messages unreliable?  Aren't we sending messages over tcp?

**Comment by j0sh** [comment]
We don't know if directed messages actually reach the intended destination after sending them. There could be partitions within the libp2p network, the destination node could be experiencing issues, and there is a large element of trust in the peers you're connected to (eclipse attacks etc).

**Comment by ericxtang** [comment]
Ahh.  The current implementation of `ping` is only to direct neighbors.  I see it just as a simple debugging tool to see if the node still has connectivity to a specific neighbor.  Added the nonce stuff though - we can use it in the future.

File: basic_out_stream.go (hunk 1 above)
Comment (j0sh) [reply basic_out_stream.go:95]: OK if we don't have any other behavior that assumes `ErrOutStream` here, similar to how we have special expectations for `ErrHandleMsg`

File: data.go [hunk 3]
```diff
@@ -81,6 +81,9 @@ type NodeStatusDataMsg struct {
 	NotFound bool
 }
 
+type PingDataMsg string
+type PongDataMsg string
```
Comment (j0sh) [inline data.go:85]: I was actually thinking more of a struct here so we could have the option to ignore later / unknown fields, but I don't think the current scheme is flexible enough for that, and it might all be moot anyway depending on what we decide for https://github.com/livepeer/go-livepeer-basicnet/issues/33 . All that being said, this is OK right now.

**Comment by j0sh** [comment]
LGTM 🚢

**Review by j0sh — APPROVED** [review]

**Comment by ghost** [comment]
@j0sh The noncing idea is great idea. 
It means we end up with a graph for every user node and for each edge off a user we know its latest IP and the DateTime changed. 
Thats really useful for mesh logic later .
